engine/crates/composition/src/validate.rs

```diff
@@ -4,9 +4,22 @@ type ValidateContext<'a> = crate::ComposeContext<'a>;
 
 /// Pre-composition validations happen here.
 pub(crate) fn validate(ctx: &mut ValidateContext<'_>) {
+    validate_query_nonempty(ctx);
     validate_selections(ctx);
 }
 
+fn validate_query_nonempty(ctx: &mut ValidateContext<'_>) {
+    if ctx
+        .subgraphs
+        .iter_subgraphs()
+        .filter_map(|subgraph| subgraph.query_type())
+        .all(|query_type| query_type.fields().next().is_none())
+    {
+        ctx.diagnostics
+            .push_fatal(String::from("None of the subgraphs defines root query fields."));
+    }
+}
```

Comment:
We have that validation in graphql-schema-validation, I think. Double checking.

Reply:
https://github.com/grafbase/grafbase/blob/main/engine/crates/validation/tests/validation_errors/objects_nonempty.errors.txt this is the one I was thinking of, but we're not testing `type Something` without curly braces, so I'll make sure we reject that too. That code path is not taken if the schemas aren't validated before composition / in from_sdl() though. Maybe we could validate as a first step inside from_sdl(), but I'm a bit worried about breakage.

Reply:
Indeed we weren't validating against `type Test` (without braces) properly, so I added that and pushed. That leaves open the question of whether we should re-validate in composition.

Reply:
> That leaves open the question of whether we should re-validate in composition.

Yeah....

It definitely would have made this issue less subtle, so I am for it in that sense.  What kind of breakage are you worried about?

Reply:
That we have tests using schemas that wouldn't pass validation. But the correct thing to do would be fixing them. Let's unblock using `main` in grafbase/api and do this as a next PR.

Reply:
A worry I have is that `grafbase/api` still has [bad schemas in it](https://github.com/search?q=repo%3Agrafbase%2Fapi%20%22type%20Query%22&type=code) but the tests where you pulled this brach in aren't failing.

Reply:
I did the same search locally, and I think the two `type Query`s without fields are in code branches that aren't taken in tests. I haven't dug more to see if we could remove / replace these, but I will before submitting the api PR to upgrade the deps.

Reply:
The EMPTY_INITAL_SCHEMA looks like it's only used for diffs, we should replace that with an Option.

Reply:
Good question, I'm perplexed now.

Reply:
Maybe we have schemas that don't define Query at all (not even `type Query`)?

Reply:
(checking)

Reply:
`get_or_create_federated_graph` returns `None` for new graphs (because its only inputs are branch slug & project ID, there's nowhere for it to _get_ a federated graph from), which hits that `unwrap_or_default("type Query")` no?

Reply:
Ok I may be confused at this point, but I think the following makes sense:

- What you said above, "I thought [this one](https://github.com/grafbase/api/blob/bea08c23c1158959b0569b9f8285182458b5688c/lambda/deployment/prepare/src/lib.rs#L173) is hit whenever you create a new federated graph, and was the cause of all the failures yesterday" is the case.
- That bad schema went through `graphql_federated_graph::from_sdl()`, causing the deployments to fail.
- With this branch, `from_sdl()` accepts schemas with empty / no Query type without crashing. That explains why the e2e tests are passing.
- If we start validating as a first step in `from_sdl()`, we should see the same tests fail again.

So it should be safe to fix the bad default schema in API, deploy the changes in this PR, then start validating. Not necessarily in this exact order, the only important part is that we start validating only after we have fixed that query. We could do it right away, but better fix main in a first step, then become strict imo.

It's a bit confusing with the different entry points so I hope this makes sense.

Reply:
(so my comment about "I did the same search locally, and I think the two type Querys without fields are in code branches that aren't taken in tests." would be wrong in this scenario)

Reply:
Just saw this comment after refreshing:

>  get_or_create_federated_graph returns None for new graphs (because its only inputs are branch slug & project ID, there's nowhere for it to get a federated graph from), which hits that unwrap_or_default("type Query") no?

yes I think so

Reply:
> So it should be safe to fix the bad default schema in API, deploy the changes in this PR, then start validating. Not necessarily in this exact order, the only important part is that we start validating only after we have fixed that query. We could do it right away, but better fix main in a first step, then become strict imo.

Yeah, cool - whatever order you want, so long as we fix all the things.